File: flutter/lib/src/interface/local_media_track.dart
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:medea_jason/src/util/rust_handles_storage.dart';

import 'track_kinds.dart';

/// Strongly referenced media track received from a
/// [`getUserMedia()`][1]/[`getDisplayMedia()`][2] request.
///
/// [1]: https://w3.org/TR/mediacapture-streams#dom-mediadevices-getusermedia
/// [2]: https://w3.org/TR/screen-capture/#dom-mediadevices-getdisplaymedia
abstract class LocalMediaTrack implements FreeableHandle {
  /// Returns the [MediaKind.Audio] if this [LocalMediaTrack] represents an
  /// audio track, or the [MediaKind.Video] if it represents a video track.
  MediaKind kind();

  /// Returns the [MediaSourceKind.Device] if this [LocalMediaTrack] is sourced
  /// from some device (webcam/microphone), or the [MediaSourceKind.Display]
  /// if it's captured via [`MediaDevices.getDisplayMedia()`][1].
  ///
  /// [1]: https://w3.org/TR/screen-capture/#dom-mediadevices-getdisplaymedia
  MediaSourceKind mediaSourceKind();

  /// Returns underlying [MediaStreamTrack] of this [LocalMediaTrack].
  MediaStreamTrack getTrack();
}

File: src/api/dart/mock/local_media_track.rs
#![allow(clippy::unused_self, missing_copy_implementations)]

use crate::{
    media::{
        track::local::LocalMediaTrack as CoreLocalMediaTrack, MediaKind,
        MediaSourceKind,
    },
    platform,
};

#[derive(Debug)]
pub struct LocalMediaTrack(pub u8);

impl From<CoreLocalMediaTrack> for LocalMediaTrack {
    fn from(_: CoreLocalMediaTrack) -> Self {
        Self(0)
    }
}

impl LocalMediaTrack {
    #[must_use]
    pub const fn kind(&self) -> MediaKind {
        MediaKind::Video
    }

    #[must_use]
    pub const fn media_source_kind(&self) -> MediaSourceKind {
        MediaSourceKind::Display
    }

    #[must_use]
    pub fn get_track(&self) -> platform::MediaStreamTrack {
        unreachable!()
    }
}
